### PR TITLE
mrpeek: (time series) plot tool

### PR DESCRIPTION
**Pull request by maxpietsch**
At the cross-hair position, key `p` produces a 2D plot of intensity values at that location across a specified dimension.

Todo:
- [x] dimension selection, here it is the current slice direction
- [x] show focus, changing the focus updates the plot
- [x] nan handling
- [x] nicer plot: add some padding

image

image

image

image

### CODE REVIEW

**Comment by jdtournier** [comment]
😍

If only we had that in `mrview` itself... 😭

**Comment by maxpietsch** [comment]
Just port ODF and fixel rendering and we can deprecate `mrview`.

**Comment by jdtournier** [comment]
😭

**Comment by jdtournier** [comment]
Not all that far fetched...
https://github.com/saitoha/SDL1.2-SIXEL/commit/ab3fccac6e34260a617be511bd8c2b2beae41952

**Comment by dchristiaens** [comment]
> Just port ODF and fixel rendering and we can deprecate `mrview`.

I would have thought these would be priority over time series plotting. Not planning on switching to fMRI are you? 😄

**Comment by jdtournier** [comment]
Looks great! I reckon this would work best with the `query_functionality` branch: `p` to prompt the question "plot along axis:", user enters axis, and that's what's displayed. Empty answer deactivates. 

I've not had a chance to check this out yet, but is the plot shown to the side of the image? And updated on change of focus?

**Comment by maxpietsch** [comment]
No no interest in fMRI time series, but in venetian blind artefacts, signal decay curves, relaxation curves, 🦓,... 

P replaces the image with the plot pane but we could use the ortho view functionality to show the slice and the plot. The plot updates with the focus, the crosshair looks different for the current focus position and is updated when modified (see first post).

**Comment by maxpietsch** [comment]
After a quite painful merge with #31, I think it's ready for prime time:

image

image

image

**Comment by jdtournier** [comment]
OK, I've just pushed some changes to have the plot show _below_ the main image, which looks like this:

Screenshot from 2020-06-23 19-42-29

This means the image retains full interactivity, so I've gone and restored all of the actions previously disabled while in plot mode. 

What do you guys reckon? 

Points to note: 

- your render might differ from mine, which I think stems from the handling of the newline character. Technically, in raw mode it shouldn't have any effect. But there's a good chance it might on some terminals. But this seems to render the same on minTTY and mlTerm on my end.

- related to the above, avoid using newline or `std::endl`, it won't do what you expect. Use `VT::move_cursor(VT::Down,1)` instead, followed by `VT::CarriageReturn`.

**Comment by jdtournier** [comment]
One more point: in due course, we should go through the code and change all writes to `std::cout` with appends to a string buffer, and condense the screen update to a single write to `std::cout` & flush. This might help to reduce some of the flicker, which is noticeable on a slower connection.

**Comment by jdtournier** [comment]
@maxpietsch: sorry, just noticed I'd nuked your _y_ axis range labels... Just fixed that, this is what it now looks like:

Screenshot from 2020-06-23 20-02-31

**Comment by maxpietsch** [comment]
Hmm looks good in the screenshots but in iTerm2, the bottom line of the plot is cut off and the plot crashes:

image

**Comment by jdtournier** [comment]
OK, did that fix it? Does it display as it should now?

**Comment by maxpietsch** [comment]
The crash might have been a botched merge on my side which I fixed in 343baeb. The line ist still cropped. I had to add another`VT::move_cursor(VT::Down,1)` before the carriage return (855b372) to get this:

image


EDIT here is mlterm:
image

**Comment by jdtournier** [comment]
That's a bummer... on my system, that extra line shift is not required, and doesn't look great. Seems iTerm2 handles the position of the cursor after a sixel image differently from mlTerm and minTTY...

**Comment by maxpietsch** [comment]
Shall we move the index text above the sixel? Not great but better than overlap or extra line ¯\_(ツ)_/¯ There would be extra space between image and plot and between plot and lower y axis label but that's better than for the main image.

**Comment by jdtournier** [comment]
Just added a quick test to see whether we can detect whether we need a newline after the sixel plot or not. The relevant test is [here](https://github.com/MRtrix3/mrpeek/pull/27/commits/8346ff4beeb4c77dc6a1543b0c5568ff82c045e8#diff-268469edc6bf7d57a39f7be7257b19aaR235). If that doesn't work, you could also try testing whether `x==1` on that line (on my system, `x` is 2 at that point, the cursor will have moved down one row after the sixel dump).

**Comment by jdtournier** [comment]
OK, I've just modified the test to actually check the row number - seems more direct somehow given the issue. 
I've also moved all the handling of the issue within the sixel encoder, so it shouldn't need special handling outside of that - the results should match regardless of terminal. :crossed_fingers:

File: src/sixel.h [hunk 1]
```diff
@@ -173,6 +177,8 @@ namespace MR {
           }
           // replace last character from $ (carriage return) to '-' (newline):
           out.back() = '-';
+          if (need_newline_after_sixel)
+            out += VT::move_cursor (VT::Down,1) + VT::CarriageReturn;
```
Comment (maxpietsch) [inline src/sixel.h]: this line breaks sixels in iTerm2, I get text to the terminal instead of the image

Comment (jdtournier) [reply src/sixel.h]: Sorry, my mistake... But it looks like you've already figured it out?

**Comment by maxpietsch** [comment]
Works:

iTerm2
image

mlTerm
image

**Comment by jdtournier** [comment]
:tada:

**Review by jdtournier — APPROVED** [review]

**Comment by jdtournier** [comment]
Good to go? Feel free to merge!

**Comment by maxpietsch** [comment]
🍾

**Pull request merged**